File: BackEnd_JS/web_crawler/lazada_crawler.js
import puppeteer from "puppeteer";

const config = {
  url: "https://www.lazada.vn/catalog/?q=b%E1%BB%99t+m%C3%AC",
  product_selector: 'div[data-qa-locator="product-item"]',
  name_selector: ".RfADt a",
  price_selector: ".aBrP0 .ooOxS",
  next_button_selector: "li.ant-pagination-next a",
  max_pages: 2,
};

export async function lazadaCrawlSite(config) {
  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();
  
  // Thêm user-agent để tránh bị render khác
  await page.setUserAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36");

  await page.goto(config.url, { waitUntil: "networkidle2" });

  const response = [];

  // Hàm scroll mượt hơn để trigger lazy-load ảnh tốt
  async function autoScroll(page) {
  await page.evaluate(async () => {
    await new Promise((resolve) => {
      let totalHeight = 0;
      const distance = 800;
      const timer = setInterval(() => {
        const scrollHeight = document.body.scrollHeight;
        window.scrollBy(0, distance);
        totalHeight += distance;
        if (totalHeight >= scrollHeight - window.innerHeight) {
          clearInterval(timer);
          resolve();
        }
      }, 2000); // delay giữa các scroll để ảnh load
    });
  });
  await page.waitForTimeout(3000); // chờ lâu hơn để ảnh lazy load
}
  let pageIndex = 1;
  while (pageIndex <= config.max_pages) {
    await autoScroll(page); // scroll đầy đủ trang

    await page.waitForSelector(config.product_selector, { timeout: 15000 });
    const products = await page.$$(config.product_selector);

    console.log(`Trang ${pageIndex} có ${products.length} sản phẩm`);

    for (let product of products) {
      const titleElement = await product.$(config.name_selector);
      const title = titleElement
        ? await page.evaluate(el => el.innerText.trim(), titleElement)
        : "Không có tên";

      const priceElement = await product.$(config.price_selector);
      const price = priceElement
        ? await page.evaluate(el => el.innerText.trim(), priceElement)
        : "Không có giá";

      const link = await page.evaluate(el => {
        const a = el.querySelector("a");
        return a ? a.href : "Không có link";
      }, product);

      let image = "Không có ảnh";
      const imgElement = await product.$("img");
      if (imgElement) {
        const dataSrc = await page.evaluate(img => img.getAttribute("data-src"), imgElement);
        const srcset = await page.evaluate(img => img.getAttribute("srcset"), imgElement);
        const src = await page.evaluate(img => img.src, imgElement); // src thường là placeholder cuối cùng

        if (dataSrc && dataSrc.startsWith("https://")) {
          image = dataSrc;
        } else if (srcset && srcset.trim()) {
          // Lấy URL đầu tiên trong srcset (thường chất lượng tốt nhất hoặc mặc định)
          const parts = srcset.split(",");
          const firstSrc = parts[0].trim().split(" ")[0];
          if (firstSrc.startsWith("https://")) {
            image = firstSrc;
          }
        } else if (src && src.startsWith("https://") && !src.includes("placeholder") && src.length > 50) {
          // src dài và không chứa placeholder mới là ảnh thật
          image = src;
        }
      }

      response.push({ title, price, link, image });
    }

    // Chuyển trang
    const nextButton = await page.$(config.next_button_selector);
    if (!nextButton || pageIndex >= config.max_pages) break;

    await Promise.all([
      nextButton.click(),
      page.waitForNavigation({ waitUntil: "networkidle0" }),
    ]);
    await page.waitForTimeout(1000);
    pageIndex++;
  }

  await browser.close();
  console.log(JSON.stringify(response, null, 2));
  return response.filter(item => {
    return item.image != "Không có ảnh"
  });
}

// run test
lazadaCrawlSite(config);

File: BackendPython/app/crawler/lazada_crawler.py
import asyncio
import json
import re
import sys
from typing import List, Dict
from playwright.async_api import async_playwright

# UTF-8 for Windows console
if sys.platform == 'win32':
    import io
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')


# ================= CONFIG =================
MAX_PAGES = 2
SCROLL_DELAY = 1.0

PRODUCT_SELECTOR = 'div[data-qa-locator="product-item"]'
NEXT_BUTTON_SELECTOR = 'li.ant-pagination-next a'


# ================= UTILS =================
async def auto_scroll(page):
    await page.evaluate("""
        async () => {
            await new Promise(resolve => {
                let total = 0;
                const distance = 800;
                const timer = setInterval(() => {
                    window.scrollBy(0, distance);
                    total += distance;
                    if (total >= document.body.scrollHeight) {
                        clearInterval(timer);
                        resolve();
                    }
                }, 800);
            });
        }
    """)
    await page.wait_for_timeout(1500)


def normalize(text: str) -> str:
    import unicodedata
    text = unicodedata.normalize("NFD", text)
    text = "".join(c for c in text if unicodedata.category(c) != "Mn")
    return text.lower().strip()


# ================= LIST CRAWL =================
async def crawl_lazada_list(page, query: str) -> List[Dict]:
    url = f"https://www.lazada.vn/catalog/?q={query}"
    await page.goto(url, wait_until="networkidle")

    products = []

    for page_index in range(1, MAX_PAGES + 1):
        print(f"\n🌐 Trang {page_index}")
        await auto_scroll(page)

        items = await page.query_selector_all(PRODUCT_SELECTOR)
        print(f"  → {len(items)} sản phẩm")

        for item in items:
            data = await item.evaluate("""
                el => {
                    const a = el.querySelector("a");
                    const title = el.querySelector(".RfADt a")?.innerText?.trim();
                    const price = el.querySelector(".aBrP0 .ooOxS")?.innerText?.trim();
                    const img =
                        el.querySelector("img")?.getAttribute("data-src") ||
                        el.querySelector("img")?.src;

                    return {
                        title,
                        price,
                        link: a?.href,
                        image: img
                    };
                }
            """)

            if data["title"] and data["link"]:
                products.append(data)

        next_btn = await page.query_selector(NEXT_BUTTON_SELECTOR)
        if not next_btn:
            break

        await asyncio.gather(
            next_btn.click(),
            page.wait_for_navigation(wait_until="networkidle")
        )

    return products


# ================= DETAIL CRAWL =================
async def crawl_lazada_detail(browser, product_url: str) -> Dict:
    page = await browser.new_page()
    await page.goto(product_url, wait_until="domcontentloaded", timeout=60000)


    detail = await page.evaluate("""
        () => {
            const uniq = arr => [...new Set(arr)];

            const result = {
                brand: "",
                sizes: [],
                colors: [],
                materials: []
            };

            // ===== BRAND =====
            const brandEl = document.querySelector(
                ".pdp-product-brand-v2__brand-link"
            );
            if (brandEl) {
                result.brand = brandEl.innerText.trim();
            }

            // ===== SIZES =====
            document
                .querySelectorAll(".sku-variable-img-name")
                .forEach(el => {
                    const size = el.innerText.trim();
                    if (size) result.sizes.push(size);
                });

            // ===== COLORS (SKU CODE) =====
            document
                .querySelectorAll(".sku-variable-name-text")
                .forEach(el => {
                    const code = el.innerText.trim();
                    if (code) result.colors.push(code);
                });

            result.sizes = uniq(result.sizes);
            result.colors = uniq(result.colors);

            return result;
        }
        """)

    await page.close()
    return detail


# ================= PIPELINE =================
async def crawl_lazada_full(query: str) -> List[Dict]:
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()

        print(f"🔍 Crawl Lazada: {query}")
        products = await crawl_lazada_list(page, query)

        print(f"\n📦 Crawl chi tiết {len(products)} sản phẩm")
        for i, product in enumerate(products, 1):
            print(f"  [{i}/{len(products)}] {product['title'][:50]}...")
            product["attributes"] = await crawl_lazada_detail(
                browser, product["link"]
            )

        await browser.close()
        return products


# ================= MAIN =================
if __name__ == "__main__":
    query = "giày nike air force 1"
    result = asyncio.run(crawl_lazada_full(query))

    print("\n================ RESULT SAMPLE ================")
    print(json.dumps(result[:2], indent=2, ensure_ascii=False))
